jupyter/d2l-java/chapter_convolutional-neural-networks/conv-layer.ipynb

```diff
@@ -0,0 +1,514 @@
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 9,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "69"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "NDArray Y = corr2d(X, K);\n",
+    "Y"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can now apply the kernel to the transposed image.\n",
+    "As expected, it vanishes. The kernel `K` only detects vertical edges."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 10,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "corr2d(X.transpose(), K);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Learning a Kernel\n",
+    "\n",
+    "Designing an edge detector by finite differences `[1, -1]` is neat\n",
+    "if we know this is precisely what we are looking for.\n",
+    "However, as we look at larger kernels,\n",
+    "and consider successive layers of convolutions,\n",
+    "it might be impossible to specify\n",
+    "precisely what each filter should be doing manually.\n",
+    "\n",
+    "Now let us see whether we can learn the kernel that generated `Y` from `X`\n",
+    "by looking at the (input, output) pairs only.\n",
+    "We first construct a convolutional layer\n",
+    "and initialize its kernel as a random array.\n",
+    "Next, in each iteration, we will use the squared error\n",
+    "to compare `Y` to the output of the convolutional layer.\n",
+    "We can then calculate the gradient to update the weight.\n",
+    "For the sake of simplicity, in this convolutional layer,\n",
+    "we will ignore the bias.\n",
+    "\n",
+    "This time, we will use the in-built `Block` and `Conv2D` class from DJL. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "83"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "Block block = Conv2D.builder()\n",
+    "                .setKernel(new Shape(1, 2))\n",
+    "                .optBias(false)\n",
+    "                .setNumFilters(1)\n",
+    "                .build();\n",
+    "\n",
+    "X = X.reshape(1,1,6,8);\n",
+    "Y = Y.reshape(1,1,6,7);\n",
+    "\n",
+    "Loss l2Loss = Loss.l2Loss();\n",
+    "\n",
+    "LearningRateTracker lrt = LearningRateTracker.fixedLearningRate(0.4f);\n",
+    "Optimizer sgd = Optimizer.sgd().setLearningRateTracker(lrt).build();\n",
+    "\n",
+    "TrainingConfig config = new DefaultTrainingConfig(Loss.l2Loss()).optOptimizer(sgd);\n",
+    "Model model = Model.newInstance(\"conv2D\");\n",
+    "model.setBlock(block);\n",
+    "Trainer trainer = model.newTrainer(config);\n",
+    "trainer.initialize(X.getShape());\n",
+    "\n",
+    "NDArray lossVal = null;\n",
+    "for(int i=0; i < 10; i++){\n",
+    "    \n",
+    "     try (GradientCollector gc = Engine.getInstance().newGradientCollector()) {\n",
```

Comment:
This example seems to be half using Trainer and half from scratch. I think either way would be fine, but be consistent with the choice